File: packages/server/src/raw.ts
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: Copyright contributors to the Slidra project

import type { ServerResponse } from "node:http";
import { SlidraError, SlidraNotFoundError } from "./slidra/errors.js";
import { readPresentationBytes } from "./slidra/reads.js";
import { MEDIA_MIME_TYPES } from "./media-types.js";

/**
 * Byte-preserving read for `assets/` content (images, video, audio) that
 * the browser needs as its original bytes, not the decoded text `cat`
 * returns (ticket #11). Kept as its own module, separate from serve.ts's
 * command-dispatch flow, because this read deliberately does not go
 * through `registry.dispatch` — see the comment at its one call site in
 * serve.ts for why.
 */

// Media extensions/MIME types come from `media-types.ts` (NOOP-90/T4,
// [E4.T9]/F7) — a literal expansion of `packages/core`'s `MEDIA_FORMATS`
// table (the server no longer imports that package at all). This table
// must stay in lockstep with the player's own extension allow-list
// (packages/web/src/player-plan.ts's VIDEO_EXTENSIONS/AUDIO_EXTENSIONS,
// ticket #30) — an extension the player accepts but this table does not
// falls back to application/octet-stream below, which some browsers refuse
// to decode as media even though the bytes are fine. A web test
// (packages/web/test/player-plan.test.ts) asserts every entry in the
// player's allow-list resolves to a non-octet-stream type here,
// specifically to keep the two lists from drifting apart again.

const MIME_TYPES: Record<string, string> = {
  ...MEDIA_MIME_TYPES,
  ".svg": "image/svg+xml",
  ".json": "application/json",
  // Fonts embedded under fonts/ (ticket #71) — served to the sandboxed
  // srcdoc iframe's @font-face over /api/raw/, same as any other asset.
  ".ttf": "font/ttf",
  ".otf": "font/otf",
  ".txt": "text/plain; charset=utf-8",
};

/**
 * Derives a Content-Type from the virtual path's extension only — no
 * content sniffing, no MIME library. An unrecognised extension is served
 * as `application/octet-stream`, never guessed.
 */
export function rawContentTypeFor(virtualPath: string): string {
  const dot = virtualPath.lastIndexOf(".");
  if (dot === -1) {
    return "application/octet-stream";
  }
  const extension = virtualPath.slice(dot).toLowerCase();
  return MIME_TYPES[extension] ?? "application/octet-stream";
}

/**
 * One resolved byte range, or an explicit verdict about why no single
 * range could be produced.
 *
 * `ignore` is RFC 9110's "a recipient that does not understand the Range
 * header field, or the range unit, MUST ignore it and serve the whole
 * representation" — a syntactically broken header is not an error, it is a
 * plain 200. `unsatisfiable` is the opposite: the client asked in a form we
 * understand, and the answer is genuinely "no such bytes", which owes a 416
 * rather than a silent full body.
 */
type RangeResolution =
  | { kind: "ignore" }
  | { kind: "satisfiable"; start: number; end: number }
  | { kind: "unsatisfiable"; reason: string };

/**
 * Resolves a `Range` header against a known total size. Only single `bytes=`
 * ranges are supported: multi-range would require a multipart/byteranges
 * body, and answering it with just the first part would silently hand the
 * client different bytes than it asked for, so it is refused outright.
 */
export function resolveByteRange(rangeHeader: string | undefined, totalSize: number): RangeResolution {
  if (rangeHeader === undefined) {
    return { kind: "ignore" };
  }
  const match = /^bytes=(.*)$/.exec(rangeHeader.trim());
  if (match === null) {
    // Some other range unit, or not a range at all: not understood, so ignored.
    return { kind: "ignore" };
  }
  const spec = match[1].trim();
  if (spec.includes(",")) {
    return { kind: "unsatisfiable", reason: "Multi-range requests are not supported, only one byte range may be requested at a time" };
  }
  const parts = /^(\d*)-(\d*)$/.exec(spec);
  if (parts === null || (parts[1] === "" && parts[2] === "")) {
    return { kind: "ignore" };
  }
  const [, rawStart, rawEnd] = parts;

  if (rawStart === "") {
    // Suffix range: the last N bytes.
    const suffixLength = Number(rawEnd);
    if (suffixLength === 0 || totalSize === 0) {
      return { kind: "unsatisfiable", reason: "Requested byte range is out of bounds" };
    }
    const start = Math.max(0, totalSize - suffixLength);
    return { kind: "satisfiable", start, end: totalSize - 1 };
  }

  const start = Number(rawStart);
  if (start >= totalSize) {
    return { kind: "unsatisfiable", reason: "Requested byte range is out of bounds" };
  }
  // An absent or over-long end is clamped to the last byte.
  const end = rawEnd === "" ? totalSize - 1 : Math.min(Number(rawEnd), totalSize - 1);
  if (end < start) {
    return { kind: "unsatisfiable", reason: "Requested byte range is invalid" };
  }
  return { kind: "satisfiable", start, end };
}

/**
 * Handles one `/api/raw/<virtual path>` request: looks the path up through
 * the same virtual tree every other read uses, and writes back the file's
 * exact bytes with a derived Content-Type and Content-Length.
 *
 * `readPresentationFileBytes` can fail for reasons that must not be
 * collapsed into the same response, so the classification here is
 * deliberately the narrow way round: only `SlidraNotFoundError` (a
 * `SlidraError` subtype that positively means "the virtual path does
 * not resolve to a file — not found, or resolves to a directory, or the
 * presentation id itself is unknown") is a 404. Every other
 * `SlidraError` — an I/O failure reading the file or an earlier
 * directory, a corrupt registry, or any subtype added later that nobody
 * has taught this function about yet — is a 500, because "not an
 * instance of SlidraNotFoundError" is not evidence that the asset is
 * missing. This is the inverse of an earlier version that treated
 * "not an I/O error" as proof of absence, which silently 404'd every new
 * error kind until someone noticed (ticket #11, fourth fix round). Either
 * way the response body is `error.message`, which — like every
 * `SlidraError` message — never contains the real filesystem path
 * (ADR-0004); only the caller-supplied virtual path may appear in it.
 *
 * `Range` handling (ticket #13) deliberately happens *after* the bytes have
 * been read successfully, so it cannot disturb that classification: a range
 * request against a missing asset is still a 404, not a 416. The whole file
 * is read into memory either way — `readPresentationFileBytes` returns a
 * Buffer and the range is a slice of it. That is enough for what the browser
 * needs from a range request (an exact 206 window, which is what Safari and
 * seeking depend on); streaming the file off disk is a separate concern.
 */
export async function handleRawRequest(
  presentationId: string,
  virtualPath: string,
  res: ServerResponse,
  rangeHeader: string | undefined,
): Promise<void> {
  let bytes: Buffer;
  try {
    bytes = await readPresentationBytes(presentationId, virtualPath);
  } catch (error) {
    if (error instanceof SlidraNotFoundError) {
      res.writeHead(404, { "Content-Type": "application/json; charset=utf-8" });
      res.end(JSON.stringify({ error: error.message }));
      return;
    }
    if (error instanceof SlidraError) {
      res.writeHead(500, { "Content-Type": "application/json; charset=utf-8" });
      res.end(JSON.stringify({ error: error.message }));
      return;
    }
    throw error;
  }

  const contentType = rawContentTypeFor(virtualPath);
  const range = resolveByteRange(rangeHeader, bytes.length);
  if (range.kind === "unsatisfiable") {
    res.writeHead(416, {
      "Content-Type": "application/json; charset=utf-8",
      "Content-Range": `bytes */${bytes.length}`,
      "Accept-Ranges": "bytes",
    });
    res.end(JSON.stringify({ error: range.reason }));
    return;
  }
  if (range.kind === "satisfiable") {
    const slice = bytes.subarray(range.start, range.end + 1);
    res.writeHead(206, {
      "Content-Type": contentType,
      "Content-Length": slice.length,
      "Content-Range": `bytes ${range.start}-${range.end}/${bytes.length}`,
      "Accept-Ranges": "bytes",
      "Cache-Control": "no-store",
    });
    res.end(slice);
    return;
  }

  res.writeHead(200, {
    "Content-Type": contentType,
    "Content-Length": bytes.length,
    // The virtual path is not a content-addressed URL: the same
    // `/api/raw/<path>` can legitimately serve different bytes over time,
    // since the file on disk — not the URL — is the source of truth
    // (live reload exists precisely because the file can change under the
    // browser's feet). No response carries an ETag or Last-Modified for the
    // browser to revalidate against, so caching is not merely stale-prone
    // here, it is unconditionally wrong: treat every response as one-shot.
    "Cache-Control": "no-store",
    // Without this the browser has no reason to believe a Range request
    // would be honoured, and downloads large video in full before playing.
    "Accept-Ranges": "bytes",
  });
  res.end(bytes);
}

File: crates/slidra/src/server/raw.rs
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: Copyright contributors to the Slidra project

//! `GET /raw/<virtual-path>` ([S11.F9], #404 Scope) — byte-preserving
//! asset serving with HTTP Range support, in-process port of
//! `packages/server/src/raw.ts`. Behavior contract (Plan §4's table, and
//! `raw.ts`'s own `resolveByteRange`/`handleRawRequest`):
//!
//! - A missing asset is 404 EVEN for a Range request — the read happens
//!   before Range is ever inspected (`raw.ts:148-`), so a Range against a
//!   missing file is never 416.
//! - `Range` header absent, unparseable, multi-range, or naming an unknown
//!   unit -> the whole file, 200, `Accept-Ranges: bytes`.
//! - A satisfiable single range -> 206 with `Content-Range`.
//! - An unsatisfiable range (suffix of a 0-byte file, `start >= total`,
//!   `end < start`) -> 416 with `Content-Range: bytes */<len>`.
//! - Every response carries `Accept-Ranges: bytes`; 200/206 additionally
//!   carry `Cache-Control: no-store` (no ETag/Last-Modified — every
//!   response is one-shot, never conditional).

use std::net::TcpStream;

use crate::server::credential::CallerKind;
use crate::server::{self, RawRequest};
use crate::workspace::virtual_fs;

const READ_CALLERS: &[CallerKind] = &[CallerKind::Editor, CallerKind::Viewer];

pub(crate) fn handle(request: &RawRequest, stream: &mut TcpStream, raw_virtual_path: &str) {
    let Some(virtual_path) = server::percent_decode(raw_virtual_path) else {
        server::write_json_error(stream, 400, "invalid path encoding");
        return;
    };
    let Some((_credential, work_dir)) =
        server::authorize_deck_scoped(request, stream, READ_CALLERS)
    else {
        return;
    };
    let bytes = match virtual_fs::read_virtual_file_bytes(&work_dir, &virtual_path) {
        Ok(bytes) => bytes,
        Err(err) => {
            let status = match &err {
                crate::errors::SlidraError::NotFound(_) => 404,
                crate::errors::SlidraError::InvalidRequest(_) => 500,
            };
            server::write_json_error(stream, status, err.message());
            return;
        }
    };
    let content_type = content_type_for(&virtual_path);
    let total = bytes.len();
    match resolve_byte_range(request.header("range"), total) {
        RangeOutcome::Ignore => {
            server::write_body_response(
                stream,
                200,
                content_type,
                &[
                    ("Cache-Control", "no-store".to_string()),
                    ("Accept-Ranges", "bytes".to_string()),
                ],
                &bytes,
            );
        }
        RangeOutcome::Satisfiable { start, end } => {
            let slice = &bytes[start..=end];
            server::write_body_response(
                stream,
                206,
                content_type,
                &[
                    ("Content-Range", format!("bytes {start}-{end}/{total}")),
                    ("Accept-Ranges", "bytes".to_string()),
                    ("Cache-Control", "no-store".to_string()),
                ],
                slice,
            );
        }
        RangeOutcome::Unsatisfiable { reason } => {
            let body = serde_json::json!({ "error": reason }).to_string();
            server::write_body_response(
                stream,
                416,
                "application/json; charset=utf-8",
                &[
                    ("Content-Range", format!("bytes */{total}")),
                    ("Accept-Ranges", "bytes".to_string()),
                ],
                body.as_bytes(),
            );
        }
    }
}

#[derive(Debug, PartialEq, Eq)]
enum RangeOutcome {
    Ignore,
    Satisfiable { start: usize, end: usize },
    Unsatisfiable { reason: String },
}

/// Ported from `raw.ts:77-116`'s `resolveByteRange`. `range_header` is the
/// raw `Range` header value (case-insensitively looked up by the caller),
/// `total_size` the full asset length.
fn resolve_byte_range(range_header: Option<&str>, total_size: usize) -> RangeOutcome {
    let Some(header) = range_header else {
        return RangeOutcome::Ignore;
    };
    let Some(spec) = header.strip_prefix("bytes=") else {
        return RangeOutcome::Ignore;
    };
    if spec.contains(',') {
        return RangeOutcome::Unsatisfiable {
            reason: "multi-range requests are not supported".to_string(),
        };
    }
    let Some((start_raw, end_raw)) = spec.split_once('-') else {
        return RangeOutcome::Ignore;
    };
    if start_raw.is_empty() && end_raw.is_empty() {
        return RangeOutcome::Ignore;
    }
    if !start_raw.chars().all(|c| c.is_ascii_digit())
        || !end_raw.chars().all(|c| c.is_ascii_digit())
    {
        return RangeOutcome::Ignore;
    }

    if start_raw.is_empty() {
        // Suffix range: the last `N` bytes.
        let Ok(suffix_length) = end_raw.parse::<u64>() else {
            return RangeOutcome::Ignore;
        };
        if suffix_length == 0 || total_size == 0 {
            return RangeOutcome::Unsatisfiable {
                reason: "range not satisfiable".to_string(),
            };
        }
        let total = total_size as u64;
        let start = total.saturating_sub(suffix_length);
        return RangeOutcome::Satisfiable {
            start: start as usize,
            end: (total - 1) as usize,
        };
    }

    let Ok(start) = start_raw.parse::<u64>() else {
        return RangeOutcome::Ignore;
    };
    if start >= total_size as u64 {
        return RangeOutcome::Unsatisfiable {
            reason: "range not satisfiable".to_string(),
        };
    }
    let end = if end_raw.is_empty() {
        total_size as u64 - 1
    } else {
        match end_raw.parse::<u64>() {
            Ok(raw_end) => raw_end.min(total_size as u64 - 1),
            Err(_) => return RangeOutcome::Ignore,
        }
    };
    if end < start {
        return RangeOutcome::Unsatisfiable {
            reason: "range not satisfiable".to_string(),
        };
    }
    RangeOutcome::Satisfiable {
        start: start as usize,
        end: end as usize,
    }
}

/// `raw.ts:30-53`'s `rawContentTypeFor` — extension-only lookup, no
/// sniffing; unknown extension -> `application/octet-stream`.
fn content_type_for(virtual_path: &str) -> &'static str {
    let ext = virtual_path
        .rfind('.')
        .map(|i| virtual_path[i + 1..].to_ascii_lowercase())
        .unwrap_or_default();
    match ext.as_str() {
        "svg" => "image/svg+xml",
        "json" => "application/json",
        "ttf" => "font/ttf",
        "otf" => "font/otf",
        "txt" => "text/plain; charset=utf-8",
        "png" => "image/png",
        "jpg" | "jpeg" => "image/jpeg",
        "gif" => "image/gif",
        "webp" => "image/webp",
        "mp4" => "video/mp4",
        "webm" => "video/webm",
        "mp3" => "audio/mpeg",
        "wav" => "audio/wav",
        "pdf" => "application/pdf",
        "csv" => "text/csv; charset=utf-8",
        "woff" => "font/woff",
        "woff2" => "font/woff2",
        _ => "application/octet-stream",
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn no_range_header_ignores() {
        assert_eq!(resolve_byte_range(None, 100), RangeOutcome::Ignore);
    }

    #[test]
    fn unknown_unit_ignores() {
        assert_eq!(
            resolve_byte_range(Some("items=0-1"), 100),
            RangeOutcome::Ignore
        );
    }

    #[test]
    fn multi_range_is_unsatisfiable() {
        assert_eq!(
            resolve_byte_range(Some("bytes=0-1,2-3"), 100),
            RangeOutcome::Unsatisfiable {
                reason: "multi-range requests are not supported".to_string()
            }
        );
    }

    #[test]
    fn both_halves_empty_ignores() {
        assert_eq!(
            resolve_byte_range(Some("bytes=-"), 100),
            RangeOutcome::Ignore
        );
    }

    #[test]
    fn suffix_range_of_zero_length_file_is_unsatisfiable() {
        assert_eq!(
            resolve_byte_range(Some("bytes=-10"), 0),
            RangeOutcome::Unsatisfiable {
                reason: "range not satisfiable".to_string()
            }
        );
    }

    #[test]
    fn suffix_zero_is_unsatisfiable() {
        assert_eq!(
            resolve_byte_range(Some("bytes=-0"), 100),
            RangeOutcome::Unsatisfiable {
                reason: "range not satisfiable".to_string()
            }
        );
    }

    #[test]
    fn suffix_range_clamps_to_start_of_file() {
        assert_eq!(
            resolve_byte_range(Some("bytes=-1000"), 100),
            RangeOutcome::Satisfiable { start: 0, end: 99 }
        );
    }

    #[test]
    fn start_beyond_total_is_unsatisfiable() {
        assert_eq!(
            resolve_byte_range(Some("bytes=100-"), 100),
            RangeOutcome::Unsatisfiable {
                reason: "range not satisfiable".to_string()
            }
        );
    }

    #[test]
    fn open_ended_range_reads_to_the_end() {
        assert_eq!(
            resolve_byte_range(Some("bytes=50-"), 100),
            RangeOutcome::Satisfiable { start: 50, end: 99 }
        );
    }

    #[test]
    fn end_is_clamped_never_an_error() {
        assert_eq!(
            resolve_byte_range(Some("bytes=0-99999"), 100),
            RangeOutcome::Satisfiable { start: 0, end: 99 }
        );
    }

    #[test]
    fn end_before_start_is_unsatisfiable() {
        assert_eq!(
            resolve_byte_range(Some("bytes=50-10"), 100),
            RangeOutcome::Unsatisfiable {
                reason: "range not satisfiable".to_string()
            }
        );
    }

    #[test]
    fn ordinary_middle_range_is_satisfiable() {
        assert_eq!(
            resolve_byte_range(Some("bytes=10-19"), 100),
            RangeOutcome::Satisfiable { start: 10, end: 19 }
        );
    }

    #[test]
    fn content_type_falls_back_to_octet_stream() {
        assert_eq!(
            content_type_for("assets/data.bin"),
            "application/octet-stream"
        );
        assert_eq!(content_type_for("assets/pic.PNG"), "image/png");
        assert_eq!(content_type_for("assets/img.svg"), "image/svg+xml");
    }
}
